src/charm.py

```diff
@@ -155,23 +159,166 @@ def _on_munge_available(self, event):
-        self.set_db_info(
-            {
-                "db_username": event.username,
-                "db_password": event.password,
-                "db_hostname": "127.0.0.1",
-                "db_port": "3306",
-                "db_name": "slurm_acct_db",
-            }
-        )
+
+        socket_endpoints = []
+        tcp_endpoints = []
+        if not event.endpoints:
+            # This is 100% an error condition that the charm doesn't know how to handle
+            # and is an unexpected condition. Raise an error here to fail the hook in
+            # a bad way. The event isn't deferred as this is a situation that requires
+            # a human to look at and resolve the proper next steps. Reprocessing the
+            # deferred event will only result in continual errors.
+            logger.error(f"No endpoints provided: {event.endpoints}")
+            self.unit.status = ErrorStatus("No database endpoints")
+            raise ValueError(f"Unexpected endpoint types: {event.endpoints}")
+
+        for endpoint in [ep.strip() for ep in event.endpoints.split(",")]:
+            if not endpoint:
+                continue
+
+            if endpoint.startswith("file://"):
+                socket_endpoints.append(endpoint)
+            else:
+                tcp_endpoints.append(endpoint)
+
+        db_info = {
+            "db_username": event.username,
+            "db_password": event.password,
+            "db_name": SLURM_ACCT_DB,
+        }
+
+        if socket_endpoints:
+            # Socket endpoints will be preferred. This is the case when the mysql
+            # configuration is using the mysql-router on the local node.
+            logger.debug("Updating environment for mysql socket access")
+            if len(socket_endpoints) > 1:
+                logger.warning(
+                    f"{len(socket_endpoints)} socket endpoints are specified, "
+                    f"but only first one will be used."
+                )
+            # Make sure to strip the file:// off the front of the first endpoint
+            # otherwise slurmdbd will not be able to connect to the database
+            socket = socket_endpoints[0][7:]
+            self._update_defaults(mysql_unix_port=f'"{socket}"')
+        elif tcp_endpoints:
+            # This must be using TCP endpoint and the connection information will
+            # be host_address:port. Only one remote mysql service will be configured
+            # in this case.
+            logger.debug("Using tcp endpoints specified in the relation.")
+            if len(tcp_endpoints) > 1:
+                logger.warning(
+                    f"{len(tcp_endpoints)} tcp endpoints are specified, "
+                    f"but only the first one will be used."
+                )
+            addr, port = tcp_endpoints[0].rsplit(":", 1)
+            # Check IPv6 and strip any brackets
+            if addr.startswith("[") and addr.endswith("]"):
+                addr = addr[1:-1]
```

Comment:
I did this based on a search of the [Slurm code](https://github.com/search?q=repo%3ASchedMD%2Fslurm%20StorageHost&type=code) combined with the [MySQL documentation on IPv6](https://dev.mysql.com/doc/refman/8.0/en/ipv6-remote-connections.html), with all IPv6 values having no brackets. I believe it to be correct, however - I think full IPv6 validation should be done separately. Since I was touching this and it could be IPv6, I made this work as defined in the MySQL bits.